pkg/csi/service/vanilla/controller.go

```diff
@@ -1137,15 +1139,121 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("CreateSnapshot: called with args %+v", *req)
-	return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+
+	isBlockVolumeSnapshotEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+	}
+
+	// Validate arguments
+	volumeID := req.GetSourceVolumeId()
+	if len(volumeID) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"CreateSnapshot Source Volume ID must be provided")
+	}
+
+	if len(req.Name) == 0 {
+		return nil, logger.LogNewErrorCode(log, codes.InvalidArgument,
+			"Snapshot name must be provided")
+	}
+
+	// Check if volume already exists
+	volumeIds := []cnstypes.CnsVolumeId{{Id: volumeID}}
+	queryFilter := cnstypes.CnsQueryFilter{
+		VolumeIds: volumeIds,
+	}
+	queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"The volume %q to be snapshotted is not available: %v", volumeID, err)
+	}
+
```

Comment:
I believe there needs to be special handling when volume migration is enabled, could you check on this with @divyenpatel

Reply:
Discussed with Divyen about the background of CSI migration. I think we can just block any snapshot on migrated vSphere volume for now. To support it, we need more thorough thinking and discussion on this case. cc: @xing-yang